added report back (#1)
rasmus-kirk/halo-accumulation: report/report.md

---
title: Investigating IVC with Halo2
date: \today
author: 
  - Rasmus Kirk Jakobsen - 201907084
geometry: margin=2cm
---

\tableofcontents
\newpage

# Introduction

Halo2, can be broken down into the following components:

- **Plonk**: A general-purpose, potentially zero-knowledge, proof scheme.
- **$\PCDL$**: A Polynomial Commitment Scheme in the Discrete Log setting.
- **$\ASDL$**: An Accumulation Scheme in the Discrete Log setting.
- **Pasta**: A Cycle of Elliptic Curves, namely **Pa**llas and Ve**sta**.

This project is focused on the components of $\PCDL$ and $\ASDL$. I used the

> [2020 paper](https://example.com) _"Proof-Carrying Data from Accumulation

Reviewer: ```suggestion
[2020 paper](https://eprint.iacr.org/2020/499.pdf) _"Proof-Carrying Data from Accumulation
```

Schemes"_ as a reference. The project covers both the theoretical aspects
of the scheme described in this document along with a rust implementation,
both of which can be found in the project's [repository](https://example.com).

## Prerequisites

Basic knowledge on elliptic curves, groups, interactive arguments are
assumed in the following text. There is also a heavy reliance on the Inner
Product Proof from the Bulletproofs protocol, see the following resources
on bulletproofs if need be:

- [Section 3 in the original Bulletproofs paper](https://eprint.iacr.org/2017/1066.pdf#section.3)
- [From Zero (Knowledge) to Bulletproofs writeup](https://github.com/AdamISZ/from0k2bp/blob/master/from0k2bp.pdf)
- [Rust Dalek Bulletproofs implementation notes](https://doc-internal.dalek.rs/develop/bulletproofs/notes/inner_product_proof/)
- [Section 4.1 of my bachelors thesis](https://rasmuskirk.com/documents/high-assurance-cryptography-implementing-bulletproofs-in-hacspec.pdf#subsection.4.1)

## Background and Motivation

### Proof Systems

In an Interactive Proof System we have two Interactive Turing machines

> the computationally unbounded Prover, P, and the polynomally time bounded

Reviewer: ```suggestion
the computationally unbounded Prover, P, and the polynomially time bounded
```

Verifier, V. The Prover tries to convince the Verifier of a statement $x \in L$

> language $L \subset \mathbb{B}^*$ in NP. The following properties must be true:

Reviewer: The text is a bit unclear here! What is B? Just saying L \in NP should be enough!

- Completeness: $\forall P \in ITM, x\in L \implies Pr[V_{out} = \bot] \leq \epsilon(x)$

  For all honest provers, P, where $x$ is true, the probability that the
  verifier remains unconvinced is negligible in the length of $x$.

- Soundness: $\forall P^* \in ITM, x \notin L \implies Pr[V_{out} = \top] \leq \epsilon(x)$

  For all provers, honest or otherwise, $P^*$, that tries to convince the
  verifier of a claim, $x$, that is not true, the probability that the
  verifier will be convinced is negligible in the length of $x$.

An Interactive Argument is very similar, but the honest and malicious prover
is now polynomially bounded and receives a Private Auxilliary Input, for

> example a witness:

Reviewer: I suggest removing "Private Auxiliary Input" and using "witness" everywhere. The completeness property should then be stated as something like this: for all PPT P that holds w so that (x,w) \in R, it holds that Pr[V_{out} = \bot] \leq \epsilon(x)

Author: I had the same annoyance, but was actually afraid whether "Private Auxiliary Input" necessarily meant "witness". I can assume they are always interchangeable?

Reviewer: I haven't seen the term "Private Auxiliary Input" in any ZK paper :)

Author: Same, except in the materials given to me in courses by proffessors :cry:

- Completeness: $\forall P(PAI) \in PPT, x\in L \implies Pr[V_{out} = \bot] \leq \epsilon(x)$
- Soundness: $\forall P^* \in PPT, x \notin L \implies Pr[V_{out} = \top] \leq \epsilon(x)$

Proof of knowledge is another type of Proof System, here the prover claims
to know a specific _witness_, $w$, for a statement $x$. Let $x \in L$ and
and $W(x)$ the set of witnesses for $x$ that should be accepted in the
proof. This allows us to define the following relation: $R = \{ (x,w) :
x \in L , w \in W(x) \}$

A proof of knowledge for relation $R$ with is a two party protocol $(P, V)$
with the following two properties:

- **Knowledge Completeness:** $Pr[P(w) \iff V_{out} = 1] = 1$, i.e. as in
  Interactive Proof Systems, after an interaction between the prover and
  verifier the verifier should be convinced with certainty.  
- **Knowledge Soundness:** Loosely speaking, in Knowledge Soundness we have

>   an extracter $E$, that when given a possibly malicious prover $P^*$

Reviewer: ```suggestion
  an efficient extractor $E$, that when given a possibly malicious prover $P^*$
```

  as a turing machine has at least as high as the probability that $P^*$
  convinces $V$

The above proof systems may be _zero-knowledge_, which in loose terms means
that anyone looking at the transcript, that is the interaction between
prover and verifier, will not be able to tell the difference between a real
transcript and one that is simulated. This means that an adversary will not
learn any new information, that they could not have gathered themselves. We
can define the property more formally:

- **Zero-knowledge:** $\forall V^*(\delta). \exists S_{V^*}(x) \in PPT. S_{V^*} \sim^C (P,V^*)$

> $V^*$ denotes a prover, honest or otherwise, $\d$ represents information

Reviewer: This is probably a typo: $V^*$ denotes a verifier

that $V^*$ may have from previous executions of the protocol and $(P,V^*)$

> denotes the transcript between the prover and verifier. There is three kinds

Reviewer: ```suggestion
denotes the transcript between the honest prover and (possibly) malicious verifier. There are three kinds
```

of zero-knowledge:

- **Perfect Zero-knowledge:** $\forall V^*(\delta). \exists S_{V^*}(x) \in PPT. S_{V^*} \sim^P (P,V^*)$,
  the transcripts $S_{V^*}(x)$ and $(P,V^*)$ are perfectly indistinguishable.
- **Statistical Zero-knowledge:** $\forall V^*(\delta). \exists S_{V^*}(x) \in PPT. S_{V^*} \sim^S (P,V^*)$,
  the transcripts $S_{V^*}(x)$ and $(P,V^*)$ are statistically indistinguishable.
- **Computational Zero-knowledge:** $\forall V^*(\delta). \exists S_{V^*}(x) \in PPT. S_{V^*} \sim^C (P,V^*)$,
  the transcripts $S_{V^*}(x)$ and $(P,V^*)$ are computationally
  indistinguishable, i.e. no polynomially bounded adversary $\Ac$ can
  distinguish them.

### SNARKS

SNARKs - **S**uccinct **N**on-interactive **AR**guments of **K**nowledge
- have seen increased usage due to their application in blockchains and
cryptocurrencies. They usually also function as so called general-purpose
proof schemes. This means that, given any solution to an NP-problem, it will
produce a proof that the prover knows the solution to said NP-problem. Most

> snarks also allow for zero-knowledge arguments, making them zk-SNARKs.

Reviewer: ```suggestion
SNARKs also allow for zero-knowledge arguments, making them zk-SNARKs.
```

> More concretely, imagine that Alice has todays sudoko problem $x$: She

Reviewer: ```suggestion
More concretely, imagine that Alice has todays sudoku problem $x$: She
```

claims to have a solution to this problem, her witness, $w$, and want to
convince Bob without having to send all of it. She could then use a SNARK to
create a proof to convince Bob. To do this she must first encode the sudoku
verifier as a circuit $R_x$, and then give the SNARK prover ($\SNARKProver$)

> $R_x(w,w')$ where $w'$ is a public input. This public input could correspond

Reviewer: Not sure if I understand this! What is w' and why is that necessary when public input can be part of x? We can simply write $R_x(w)$ here (unless I am missing something)

Author: It's just notation I suppose. When presented with SNARKs it was presented as above. How about:

```md
To do this she must first encode the sudoku verifier as a circuit $R_x$,
and then give the SNARK prover ($\SNARKProver$) $R_x(w)$. Any public inputs,
such as todays sudoku digit values and positions, can be included as constants 
in the circuit.
```

More explicit/clear, the most important thing is to include that public inputs are handled.

Reviewer: I think just saying that the circuit $R_x$ encodes the sudoku statement (which includes public inputs, today's sudoku values/positions, etc) should be enough

> to the known already known digits for todays sudoku. Finally she can send

Reviewer: ```suggestion
to the already known digits for todays sudoku. Finally she can send
```

that said state was computed correctly.

\begin{figure}[!H]
\centering
\begin{tikzpicture}[node distance=2cm]

    % Nodes
    \node (z0) [node] {$z_0$};
    \node (z1) [node, right=of z0] {$(z_1, \pi_1)$};
    \node (z2) [node, right=of z1] {$(z_2, \pi_2)$};
    \node (dots) [right=1cm of z2] {$\dots$};
    \node (zn) [node, right=1cm of dots] {$(z_n, \pi_n)$};

    % Arrows with labels
    \draw[arrow] (z0) -- node[above] {$F$} (z1);
    \draw[arrow] (z1) -- node[above] {$F$} (z2);
    \draw[arrow] (z2) -- node[above] {$F$} (dots);
    \draw[arrow] (dots) -- node[above] {$F$} (zn);

\end{tikzpicture}
\caption{A visualization of the relationship between $F, \vec{z}$ and $\vec{\pi}$ in an IVC setting.}
\end{figure}

The proof $\pi_i$ describes the following:

> _\textcolor{GbFg3}{"The current state $z_i$ is computed from applying $F$ $i$ times to $z_0$
> ($z_i = F^i(z_0) = F(z_{i-1})$) and the associated proof $\pi_{i-1}$ for
> the previous state is valid."}_

Or more formally, $\pi_i$ is a proof of the following claim expressed as a
circuit $R$:

$$z_i = F(z_{i-1}) \; \land \; (i = 0 \lor \exists \, \pi_{i-1}, \, \text{ s.t. } \SNARKVerifier(z_{i-1}, \pi_{i-1}) = \top)$$

> Where $V$ represents the verification circuit in the proof system we're

Reviewer: What/where is V?

Author: Yes, my bad. I changed `v` to `\SNARKVerifier` at some point for clarity. Capturing different verifiers transparently, was somewhat of a challenge, so I decided to make it more explicit.

using. This means, that we're taking the verifier, representing it as a circuit, and
then feeding it to the prover. This is not a trivial task in practice! It
also means that the verification time must be sublinear for IVC to work
properly, otherwise

**TODO**:

- Explain: Otherwise what happens? Blowup?

### Bulletproofs

> In 2016, [the Bulletproofs paper](https://eprint.iacr.org/2017/1066.pdf)

Reviewer: ```suggestion
In 2017, [the Bulletproofs paper](https://eprint.iacr.org/2017/1066.pdf)
```

was released. Bulletproofs relies on the hardness of the Discrete Logarithm
problem, and allows for an untrusted setup to generate the Common Reference
String. It has logarithmic proof size, and lends itself well efficient range
proofs. It's also possible to generate proofs for arbitrary circuits, but
with less effeciency.

At the heart of Bulletproofs lies the Inner Product Argument (IPA), wherein
a prover proves he knows two vectors, $\vec{a}, \vec{b} \in \Fb_q^n$,
with commitment $C \in \Eb(\Fb_q)$, and their corresponding inner product,
$c = \ip{\vec{a}}{\vec{b}}$. It creates this non-interactive proof,
with only $\lg(n)$ size, by compressing the point and vectors $\lg(n)$

> times. Unfortunately, the IPA suffers, and by extension Bulletproofs, suffer

Reviewer: ```suggestion
times. Unfortunately, the IPA, and by extension Bulletproofs, suffer
```

from linear verification time, making them unsuitible for IVC.

### Accumulation Schemes

The authors of [a 2019 paper](https://eprint.iacr.org/2019/1021.pdf)
presented _Halo_ the so-called first practical example of recursive
proof composition without a trusted setup. Using a modified version
of the Bulletproofs-style Inner Product Argument (IPA), they present a
polynomial commitment scheme. Computing the evaluation of a point $z \in
\Fb_q$ on polynomial $p(X) \in \Fb^d_q[X]$ as $v = \ip{\vec{p}}{\vec{z}}$

> where $\vec{z} = (z^0, z^1, \dots, z^{d+1})$ and $\vec{p} \in \Fb^d$ is the

Reviewer: ```suggestion
where $\vec{z} = (z^0, z^1, \dots, z^{d})$ and $\vec{p} \in \Fb^{d+1}$ is the
```

coefficient vector of $p(X)$, using the IPA. However, since the the vector
$\vec{z}$ is not private, and has a certain structure, we can split the
verification algorithm in two: $\PCDLSuccinctCheck$ and $\PCDLCheck$. Using
the $\PCDLSuccinctCheck$ we can accumulate $n$ instances, and only perform

> the expensive linear check at the end of accumulation.

Reviewer: ```suggestion
the expensive linear check (i.e., $\PCDLCheck$) at the end of accumulation.
```

  if $\top = \ASDecider(acc) = \Phi(q)$, then for $\ASProver(q, acc_i)
  = acc_{i+1}$ it holds that $\top = \ASVerifier(acc_i, q, acc_{i+1}) =
  \ASDecider(acc_{i+1})$.

- **Soundness:** For all efficiently-generated accumulators $acc_i, acc_{i+1}
  \in \Acc$ and predicate inputs $q \in X$, if $\top = \ASDecider(acc_{i+1}) =
  \ASVerifier(q, acc_i, acc_{i+1})$ then, with all but negligible probability,
  $\top = \Phi(q) = \ASDecider(acc_i)$.

### IVC from Accumulation Schemes

If the $\ASVerifier$ runtime is sub-linear and we have a $\SNARKVerifierSlow$
that may run in linear time, then we can create an IVC scheme:

- $\IVCProver(z_i: \textbf{Instance}, \pi_i: \textbf{Proof}, acc_i: \Acc) \to (\textbf{Proof}, \Acc)$

  - Accumulates $(z_i, \pi_i)$ with $acc_i$ to obtain a new accumulator $acc_{i+1}$.
  - Then generates a SNARK proof $\pi_{i+1}$ of the following claim, expressed as a circuit $R$:
      $$"z_{i+1} = F(z_i) \; \land \; (i = 0 \lor \exists \, \pi_i, \, acc_i \text{ s.t. } \ASVerifier((z_i, \pi_i), acc_i, acc_{i+1}) = \top)"$$
  - Output $(\pi_{i+1}, acc_{i+1})$

- $\IVCVerifier(\pi_{i+1}: \textbf{Proof}, acc_{i+1}: \Acc) \to \Result(\top, \bot)$

>   Checks the proof: $\top \meq \SNARKVerifierSlow(\pi_{i+1}) \meq \ASDecider(acc_{i+1})$

Reviewer: SNARK verifier should take as input acc_{i+1} as well; i.e., it should be \SNARKVerifierSlow(\pi_{i+1}, acc_{i+1})

Author: From the paper:

> Given such an accumulation scheme, we can construct IVC as follows. Given a previous instance $z_i$, proof $\pi_i$, and accumulator $\text{acc}_i$, the IVC prover first accumulates $(z_i, \pi_i)$ with $\text{acc}_i$ to obtain a new accumulator $\text{acc}_{i+1}$. The IVC prover also generates a SNARK proof $\pi_{i+1}$ of the claim: "$z_{i+1} = F(z_i)$, and there exist a proof $\pi_i$ and an accumulator $\text{acc}_i$ such that the accumulation verifier accepts $((z_i, \pi_i), \text{acc}_i, \text{acc}_{i+1})$", expressed as a circuit $R$. The final IVC proof then consists of $(\pi_T, \text{acc}_T)$. _The IVC verifier checks such a proof **by running the SNARK verifier on $\pi_T$** and the accumulation scheme decider on $\text{acc}_T$._

Is the above wrong, or have I misunderstood it?

Reviewer: No, I think the above is correct; but as written, the SNARK proof $\pi_{i+1}$ is for a claim about z_{i+1}, and accumulation verifier that involves $\text{acc}{i+1}$. So, the SNARK verifier should take as input all $\pi_{i+1}$, z_{i+1}, and $\text{acc}{i+1}$

Reviewer: Also from the original paper:
![image](https://github.com/user-attachments/assets/53155fcc-9178-4267-b7ee-1df740049519)

Consider the above chain run $n$ times. As in the "simple" SNARK IVC
construction, if the SNARK verifier accepts at the end, then that means:

$$
\begin{alignedat}[b]{2}
  &\SNARKVerifierSlow(\pi_i)                               &&= \top \then \\
  &\ASVerifier((z_{n-1}, \pi_{n-1}), acc_{n-1}, acc_n)     &&= \top \then \\
  &\ASVerifier((z_{n-2}, \pi_{n-2}), acc_{n-2}, acc_{n-1}) &&= \top \then \cdots \\
  &\ASVerifier((z_0, \pi_0), acc_0, acc_1)                 &&= \top \then \\
\end{alignedat}

> $$

Author: Also just putting a pin on this, I need to recurse to z_1 instead as the verifier does not run on z_0, in my description.

Now, by the soundness property of the Accumulation Scheme, and instance
$q \in X$ will be true if $\top = \ASVerifier(q, acc_i, acc_{i+1}) =
\ASDecider(acc_{i+1})$, so if all the accumulators $\vec{acc} \in
\Acc^{n+1}$ are valid, then all the instances will be true. This is exactly
the case however due to the definition of the decider whereby checking an
accumulator $acc_i$ ensures that every previous instance is true, $\Phi(q_i)
= \top$, provided that all previous verifiers accepted.

> ### General Polynomial Commitment Schemes

Reviewer: I suggest removing "General" from the title

In the section above about SNARKs, general-purpose proof schemes were
described. Modern general-purpose (zero-knowledge) proof schemes, such as

> Sonic[^1], Plonk[^2] and Marlin[^3], commonly use PCS's _Polynomial Commitment

Reviewer: ```suggestion
Sonic[^1], Plonk[^2] and Marlin[^3], commonly use _Polynomial Commitment
```

> Schemes_ for creating their proofs. This means that different PCS's can be

Reviewer: ```suggestion
Schemes_ (PCSs) for creating their proofs. This means that different PCSs can be
```

used to get security under weaker or stronger assumptions.

<!-- AGM gives you snarks? -->
**TODO**: List the options (AGM?, BP, STARKS).

- **AGM:** Uses a trusted setup, this would give you a traditional SNARK.
- **DL:** Uses an untrusted setup, assumed secure if the Discrete Log problem is hard, the verifier is linear.
- **FRI:** Also uses an untrusted setup, assumed secure one way functions
  exist. It has a higher overhead than PCSs based on the Discrete Log assumption.

**TODO**: The assumptions on $d$ seems wrong in the code, not a big deal

A PCS allows a prover to prove to a verifier that a commited polynomial
evaluates to a certain value, $v$, given an evaluation input $z$.There are
three main functions used to prove this ($\PCSetup$ and $\PCTrim$ omitted):

- $\PCCommit(p: \Fb^{d'}_q[X], d: \Nb, \o: \Option(\Fb_q)) \to \Eb(\Fb_q)$

>   Commits to a polynomial $p$ with degree bound $d$ where $d \geq d'$ using

Reviewer: What is d'? Degree of p? Then we can perhaps say something like "...a degree-d' polynomial $p$..."?

Author: Yes exactly, it's specified in the `\Fb^{d'}_q[X]` but I like the clarification suggested.

### The Implementation

The authors also define a concrete Accumulation Scheme using the Discrete Log
assumption $\ASDL$, which uses the same algorithms as in the 2019 Halo
paper. This accumulation scheme in turn, relies heavily upon a Polynomial
Commitment Scheme, $\PCDL$, which is also described in the paper. Both of
these have been implemented as part of this project in Rust and the rest
of the document will go over these sets of algorithms, their security,
performance and implementation details.

Since these kinds of proofs can both be used for proving knowledge of a
large witness to a statement succinctly, and doing so without revealing
any information about the underlying witness, the zero-knowledgeness of the
protocol is described as _optional_. This is highlighted in the algorithmic
specifications as the parts colored \textblue{blue}. In the Rust
implementation I have chosen to include these parts as they were not too
cumbersome to implement, but since IVC is at the heart of this project,
not zero-knowledge, I have chosen to omit them from the soundness and
completeness discussions in the following sections as well as omitted the

> proofs of zero-knowledgeness.

Reviewer: ```suggestion
proofs of zero-knowledge.
```

// Since the order of the curve is prime, any non-identity point is a generator.
fn get_generator_hash(i: usize) -> PallasPoint {
    let genesis_string = "To understand recursion, one must first understand recursion"
      .as_bytes();

    // Hash `genesis_string` concatinated with `i`
    let mut hasher = Sha3_256::new();
    hasher.update(genesis_string);
    hasher.update(i.to_le_bytes());
    let hash_result = hasher.finalize();

    // Interpret the hash as a scalar field element
    let scalar = PallasScalar::from_le_bytes_mod_order(&hash_result);

    // Generate a uniformly sampled point from the uniformly sampled field element
    PallasPoint::generator() * scalar
}

/// Get public parameters
fn get_pp(n: usize) -> (PallasPoint, PallasPoint, Vec<PallasPoint>) {
    let S = get_generator_hash(0);
    let H = get_generator_hash(1);
    let mut Gs = Vec::with_capacity(n);

    for i in 2..(n + 2) {
        Gs.push(get_generator_hash(i))
    }

    (S, H, Gs)
}
```

\newpage

# $\PCDL$: The Polynomial Commitment Scheme

## Outline

We have four main functions:

- $\PCDLCommit(p: \Fb^d_q[X], \o: \Option(\Fb_q)) \to \Eb(\Fb_q)$:

>   Creates a commitment to the coefficients of the polynomial $q$ of degree

Reviewer: ```suggestion
  Creates a commitment to the coefficients of the polynomial $p$ of degree
```

$$
\begin{alignedat}[b]{1}
  \vec{L}    &= (L_1, \dots, L_{\lg(n)}) \\
  \vec{R}    &= (R_1, \dots, R_{\lg(n)}) \\
  \vec{C}    &= (C_0, \dots, C_{\lg(n)}) \\
  \vec{\xi}  &= (\xi_0, \dots, \xi_{\lg(n)}) \\
\end{alignedat}
$$

This means an honest prover will indeed produce $\vec{L}, \vec{R}$ s.t. $C_{\lg(n)} = C_0 + \sum^{\lg(n)-1}_{i=0} \xi^{-1}_{i+1} L_i + \xi_{i+1} R_i$

Let's finally look at the left-hand side of the verifying check:

\begin{align*}
  C_{\lg(n)} &= C_0 + \sum^{\lg(n)-1}_{i=0} \xi^{-1}_{i+1} L_i + \xi_{i+1} R_i && \\
             \intertext{The original definition of $C_i$:}
             &= \ip{\vec{c}_{\lg(n)}}{\vec{G}_{\lg(n)}} + \ip{\vec{c}_{\lg(n)}}{\vec{z}_{\lg(n)}} H' \\
             \intertext{Vectors have length one, so we use the single elements $c^{(0)}, G^{(0)}, c^{(0)}, z^{(0)}$ of the vectors:}
             &= c^{(0)}G^{(0)} + c^{(0)}z^{(0)} H'                                                   \\
             \intertext{The verifier has $c^{(0)} = c, G^{(0)} = U$ from $\pi \in \EvalProof$:}
             &= cU + cz^{(0)} H'                                                                     \\

>              \intertext{Then, by construction of $h(X) \in \Fb^d_q[X]$}

Author: Not sure on this, "by construction" seems incomplete, but I have a hard time proving/formalizing why this is.
